libs/synchronization/src/mutex.cpp

```diff
@@ -90,18 +90,18 @@ namespace hpx { namespace lcos { namespace local {
         HPX_ASSERT(threads::get_self_ptr() != nullptr);
 
         HPX_ITT_SYNC_RELEASING(this);
+        // Unregister lock early as the lock guard below may suspend.
+        util::unregister_lock(this);
         std::unique_lock<mutex_type> l(mtx_);
 
         threads::thread_id_type self_id = threads::get_self_id();
         if (HPX_UNLIKELY(owner_id_ != self_id))
         {
-            util::unregister_lock(this);
             HPX_THROWS_IF(ec, lock_error, "mutex::unlock",
```

Comment:
While you're looking at this. Could you please add a `l.unlock();` before the exception is thrown?

Reply:
Why are we taking a lock during a call to mutex.unlock?

Reply:
> Why are we taking a lock during a call to mutex.unlock?

To protect the internal members of the mutex itself.

Reply:
How many levels of mutex-ception are there? Do the mutexes that protect the mutexes also have mutexes protecting their internals? Could you remind me please of what internal members of the mutex need protecting?

Reply:
@biddisco `hpx::lcos::local::mutex` is implemented using our `detail::condition_variable` which requires protection.

Reply:
Yep, good catch.